nemo/collections/asr/parts/submodules/fast_rnnt_greedy_decoding.py

```diff
@@ -0,0 +1,440 @@
+                self.scores.transpose(0, 1).contiguous().reshape((self.batch_size, self.max_time, self.max_symbols)),
+                non_blocking=True,
+            )
+            self.labels_cpu.copy_(
+                self.labels.transpose(0, 1).contiguous().reshape((self.batch_size, self.max_time, self.max_symbols)),
+                non_blocking=True,
+            )
+
+            self.last_label.fill_(self.caller._SOS)
+            self.time_idx_t.fill_(0)
+
+    def __call__(
+        self,
+        x: torch.Tensor,
+        out_len: torch.Tensor,
+        device: torch.device,
+        partial_hypotheses: Optional[List[rnnt_utils.Hypothesis]] = None,
+    ):
+
+        # Need to copy x and out_len into "staging buffers", or do a graph update.
+
+        if partial_hypotheses is not None:
+            raise NotImplementedError("`partial_hypotheses` support is not supported")
+
+        assert not self.caller.preserve_alignments
+        assert not self.caller.preserve_frame_confidence
+
+        batch_size = x.shape[0]
+        # ideally we would use out_len.max() here...
+        max_time = x.shape[1]
+
+        if torch.is_autocast_enabled():
+            x = x.to(torch.get_autocast_gpu_dtype())
+
+        # What do we do if batch_size is actually smaller for the
+        # input? Is this a problem? the clone() call will fail...
+        if max_time > self.max_time or batch_size > self.batch_size:
+            self._reinitialize(max_time, batch_size, x, out_len)
+
+        torch.cuda.nvtx.range_push("Graph exec")
+        self.encoder_output[: x.shape[0], : x.shape[1], ...].copy_(x)
+        self.encoder_output_length[: out_len.shape[0]].copy_(out_len)
+        self.graph.replay()
+        torch.cuda.current_stream().synchronize()
+        torch.cuda.nvtx.range_pop()
+
+        torch.cuda.nvtx.range_push("Copy data out")
+
+        torch.cuda.nvtx.range_push("scores_cpu mask")
+        self.scores_cpu[self.labels_cpu == self.caller._blank_index] = 0.0
+        torch.cuda.nvtx.range_pop()
+        torch.cuda.nvtx.range_push("scores_cpu sum")
+        total_scores = self.scores_cpu.sum(dtype=torch.float32, axis=(1, 2))
+        torch.cuda.nvtx.range_pop()
+
+        tokens_per_timestep = (self.labels_cpu != self.caller._blank_index).sum(axis=-1)
+        torch.cuda.nvtx.range_push("repeat_interleave")
+        timesteps_packed = torch.repeat_interleave(
+            torch.arange(self.max_time).repeat(self.batch_size), tokens_per_timestep.flatten()
+        )
+        timestep_segments = tokens_per_timestep.sum(axis=-1)
+        torch.cuda.nvtx.range_pop()
+
+        torch.cuda.nvtx.range_push("labels")
+        valid_labels_mask = self.labels_cpu != self.caller._blank_index
+        labels_segments = valid_labels_mask.sum(axis=(1, 2))
+        labels_packed = self.labels_cpu[valid_labels_mask]
+        torch.cuda.nvtx.range_pop()
+
+        torch.cuda.nvtx.range_push("Convert to Hypotheses")
+        hypotheses = [
```

Comment:
Was this the section you mentioned you wanted to speedup with numba CPU ?

Reply:
Ended up doing it with pytorch, which was definitely fast enough. One tricky bit about numba was the lack of bfloat16 support. And we can end up with scores in bfloat16 if the model is running in bfloat16.